### PR TITLE
rgw: should delete in_stream_req  if conn->get_obj(...) return not zero value

### PR DESCRIPTION
**Pull request by weiqiaomiao**
Signed-off-by: weiqiaomiao wei.qiaomiao@zte.com.cn

### CODE REVIEW

**Comment by weiqiaomiao** [comment]
retest this please

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -6590,7 +6590,7 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
 {
   /* source is in a different zonegroup, copy from there */
 
-  RGWRESTStreamReadRequest *in_stream_req;
+  RGWRESTStreamReadRequest *in_stream_req = NULL;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc]: should be nullptr these days, but ok :)

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -6676,6 +6676,9 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
                       dest_mtime_weight.zone_short_id, dest_mtime_weight.pg_ver,
                       true, &cb, &in_stream_req);
   if (ret < 0) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc]: does look correct

**Comment by mattbenjamin** [comment]
@cbodley do you have any issue with this?

**Comment by cbodley** [comment]
the patch looks correct, but we might want to attack the issue in `RGWRESTConn::get_obj()` instead. if it returns an error, it should clean up the `RGWRESTStreamReadRequest` itself instead of making the caller do it

**Comment by cbodley** [comment]
i.e. at the bottom of `RGWRESTConn::get_obj()`:
```diff
-  return (*req)->get_obj(key, extra_headers, obj);
+  int r = (*req)->get_obj(key, extra_headers, obj);
+  if (r < 0) {
+    delete *req;
+    *req = nullptr;
+  }
+  return r;
```

**Comment by mattbenjamin** [comment]
@cbodley agree;  if get_obj failed, there was no reason for it to return something for us to delete
@weiqiaomiao can you revise this?

**Comment by weiqiaomiao** [comment]
@mattbenjamin  yes, i revise it.

**Comment by weiqiaomiao** [comment]
retest this please

**Comment by weiqiaomiao** [comment]
retest this please

**Comment by weiqiaomiao** [comment]
@mattbenjamin ping

**Pull request merged**